[run-webkit-tests] Support user specifying variant directly

JonWBedard opened the pull request:
#### f8b4d273d3df3adce27c788f8ee7c58d776f5921
<pre>
[run-webkit-tests] Support user specifying variant directly
<a href="https://bugs.webkit.org/show_bug.cgi?id=246228">https://bugs.webkit.org/show_bug.cgi?id=246228</a>
rdar://100907445

Reviewed by Elliott Williams.

* Tools/Scripts/webkitpy/common/find_files.py:
(_normalized_find): Consider variants of a single test file.
(_normalized_find.sorted_paths_generator):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(LayoutTestFinder._expand_variants): Handle specific variant as argument.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.test_find_glob_query):

Canonical link: <a href="https://commits.webkit.org/255305@main">https://commits.webkit.org/255305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb0d3f4a5a34ca4afc06cb006a7c87e0a655b718

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92095 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1326 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/22604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/101882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/161937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96095 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1324 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97753 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78618 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/84533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/95668 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/22604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/84533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/36145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/22604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33901 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/22604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3677 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37771 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39671 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/22604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/webkitpy/common/find_files.py:
```diff
@@ -79,7 +91,8 @@ def sort_by_directory_key(files_list):
+    paths_to_walk = itertools.chain(*(sort_by_directory_key(_filter_test_variant(path, filesystem.glob)) for path in paths))
+    all_files = itertools.chain(*(sort_by_directory_key(
+        _filter_test_variant(path, lambda p: filesystem.files_under(p, skipped_directories, file_filter))
+    ) for path in paths_to_walk))
```
Suggested refactor:

```python3
def sorted_paths_generator(fs_selector):
    for test_selector in paths_to_walk:
        path, separator, variant = test_selector.partition('?')
        matching_tests = (matching_path + separator + variant for matching_path in fs_selector(path))
        if directory_sort_key:
            yield from sorted(matching_tests, key=directory_sort_key)
        else:
            yield from matching_tests

paths_to_walk = sorted_paths_generator(filesystem.glob)
all_files = sorted_paths_generator(lambda p: filesystem.files_under(p, skipped_directories, file_filter))
```

This would replace `_filter_test_variant` and the local `sort_by_directory_key` function altogether. Using `str.partition` instead of `str.split` removes a bunch of the conditional logic.

emw-apple commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
```diff
@@ -175,6 +181,7 @@ def _real_tests(self, paths):
         # When collecting test cases, skip these directories
         skipped_directories = set(['.svn', '_svn', 'resources', 'support', 'script-tests', 'tools', 'reference', 'reftest'])
         files = find_files.find(self._port._filesystem, self._port.layout_tests_dir(), paths, skipped_directories, self._is_test_file, self._port.test_key)
+        files = [f for f in files]
```
Is this a no-op? If you're intentionally converting to a list, can you use `list(files)`?

JonWBedard replied:
Yes, will remove, leftover logging.

emw-apple commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
```diff
@@ -164,7 +169,8 @@ def _expand_variants(self, files):
                     for variant in variants:
-                        expanded.append(f + variant)
+                        if not passed_variant or variant.startswith(passed_variant):
+                            expanded.append(f + variant)
```
with the above suggestion:
```suggestion
                        if not passed_variant or variant.startswith(variant_separator + passed_variant):
                            expanded.append(f + variant)
```

emw-apple commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
```diff
@@ -148,6 +148,11 @@ def _expand_variants(self, files):
+            passed_variant = None
+            if len(split) == 2:
+                f = split[0]
+                passed_variant = '?{}'.format(split[1])
```
nit:

```suggestion
f, variant_separator, passed_variant = f.partition('?')
```

emw-apple reviewed (approved):
Seems reasonable. I tried refactoring a bit to make it easier for me to follow. If you're in a hurry, no need to take these suggestions.

JonWBedard committed:
[run-webkit-tests] Support user specifying variant directly
https://bugs.webkit.org/show_bug.cgi?id=246228
rdar://100907445

Reviewed by Elliott Williams.

* Tools/Scripts/webkitpy/common/find_files.py:
(_normalized_find): Consider variants of a single test file.
(_normalized_find.sorted_paths_generator):
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(LayoutTestFinder._expand_variants): Handle specific variant as argument.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.test_find_glob_query):

Canonical link: https://commits.webkit.org/255305@main
